peachjam/models/core_document_model.py

```diff
@@ -896,6 +896,87 @@ def save(self, *args, **kwargs):
             # first save, set the download filename
             self.set_download_filename()
 
+    def delete(self, *args, **kwargs):
+        if (
+            hasattr(self.document, "publication_file")
+            and self.document.publication_file.use_source_file
+        ):
+            raise ValidationError(
+                _(
+                    "You cannot delete this source document while 'Use source file' is checked on the publication file"
+                )
+            )
+        return super().delete(*args, **kwargs)
+
+
+class PublicationFile(AttachmentAbstractModel):
+    SAVE_FOLDER = "publication_file"
+
+    file = models.FileField(
+        _("file"), upload_to=file_location, max_length=1024, blank=True, null=True
+    )
+    document = models.OneToOneField(
+        CoreDocument,
+        related_name="publication_file",
+        on_delete=models.CASCADE,
+        verbose_name=_("document"),
+    )
+    url = models.URLField(
+        _("URL"),
+        max_length=2048,
+        null=True,
+        blank=True,
+        help_text=_("The external URL (e.g. on gazettes.africa) where this file lives"),
+    )
+    use_source_file = models.BooleanField(
+        _("use source file"),
+        default=False,
+        help_text=_(
+            "Set to True if the source file on the same document can be used instead"
+        ),
+    )
+
+    class Meta:
+        verbose_name = _("publication file")
+        verbose_name_plural = _("publication files")
+
+    def clean(self):
+        """Prefer use_source_file, then url, then file if we must.
+        Raise a ValidationError if they're all missing.
+        """
+        super().clean()
+
+        # if use_source_file was chosen, check that there is one to use
+        if self.use_source_file and hasattr(self.document, "source_file"):
+            source_file = self.document.source_file
+            self.filename = source_file.filename
+            self.mimetype = source_file.mimetype
+            self.size = source_file.size
+            self.url = None
+        elif self.use_source_file:
+            self.use_source_file = False
+
+        if self.url and not (self.filename and self.mimetype and self.size):
+            raise ValidationError(
+                _(
+                    "When you use a URL, you need to provide the filename, the mimetype, and the size too."
+                )
+            )
+
+        # final check: we need at least one of these
+        if not (self.use_source_file or self.url or self.file):
+            raise ValidationError(
+                _(
+                    "You need to provide a URL or a file, or if you checked 'Use source file' there must be one to use."
+                )
+            )
+
+    def save(self, *args, **kwargs):
+        self.clean()
```

Comment:
I suspect this may cause an issue later, but let's go with it. Usually `clean()` is only called by a view or a form that is validating inputs, so that they can be corrected. Calling it on every save() is not always what you want. But I can't think of an example right now where this could go wrong, so let's go with it.

Reply:
I'd also be happy to move all the logic back into `save()` where I had it originally — I moved the bulk into `clean()` to be able to test and demonstrate the logic in the admin form

Reply:
this immediately caused an issue 😅 when we delete the file before doing the update (or create)